### PR TITLE
Refactor AIConfig to Sanitize Input for Goal Parameters

### PR DESCRIPTION
**Pull request by pr-0f3t**
### Background
Refactor AIConfig to Sanitize Input for Goal Parameters

A user reported errors with `remove_color_codes` and some ai_settings combinations. Investigations showed that some ai_settings formats can cause goals to load as `dicts`, not strings.

Specifically, if a goal is defined as:

```yaml
- Fix the files: Download, analyze and fix dirty_data.txt
```

``` python
# It's interpreted as a `dict`

goal = {"Fix the files": "Download, analyze and fix dirty_data.txt"}

# Not a string:

goal = "Fix the files: Download, analyze and fix dirty_data.txt"
```

### Changes
Details:
- Modified `ai_config.py` to correctly handle and sanitize user input for AI goals and convert them to formatted strings, to fix an issue where some specially formatted ai_settings.yaml files were causing goals to load as list[dict]
- `test_ai_config.py` includes a test for the `sanitize_input` function in `AIConfig` class.

### Documentation
No documentation change needed

### Test Plan
Tests were updated

### PR Quality Checklist
- [ x] My pull request is atomic and focuses on a single change.
- [ x] I have thoroughly tested my changes with multiple different prompts.
- [ x] I have considered potential risks and mitigations for my changes.
- [ x] I have documented my changes clearly and comprehensively.
- [ x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: autogpt/config/ai_config.py [hunk 1]
```diff
@@ -79,11 +79,31 @@ def load(config_file: str = SAVE_FILE) -> "AIConfig":
...
+        Return goals as list of strings, in the format that the user intended,
+        fixing some cases where the goals are loaded from yaml as a list of dicts.
+
+        Parameters:
+            goals: The goals to sanitize (can be a string, dict, or other object)
+
+        Returns:
+            A list of strings, where each string is a goal.
+        """
+        to_str = (
```
Comment (collijk) [inline autogpt/config/ai_config.py]:
```suggestion
        goals = [str(goal).strip("{}").replace("'", "") if isinstance(goal, dict) else str(goal) for goal in goals]
```

Comment (collijk) [reply autogpt/config/ai_config.py]: Like generally, there are some non-pythonic things here.  These are style things, and so not critical, but an explanation: 

- You don't typically want to assign lambda functions to variables.  The point of a lambda is that it's anonymous.  You should use the def syntax if you want to name a function. This is explicitly mentioned in pep8: https://peps.python.org/pep-0008/#programming-recommendations
- functional programming, while supported, isn't really first class in python.  You should strongly prefer list comprehensions over functional applications of map, reduce, filter, etc.  See https://www.artima.com/weblogs/viewpost.jsp?thread=98196 for some context.

Comment (lc0rp) [reply autogpt/config/ai_config.py]: Sigh, there aren't many places where one gets to use both lambda and map in py 3 :)

I had implemented this PR as an inline list comprehension, but  it looked barely readable (I'd kept the k: v for k, v in... syntax), and for some reason I'd recalled (wrongly) that map() was faster. 

The strip("{}").replace() fixes readability. I'll probably just put it back inline then. The separate function seems redundant, unless you object.

Does the strip().replace() syntax have edge cases worth considering? I can think of, maybe, quotes in the string, but none where losing the " ' " will be catastrophic... 

"Goal: critique the first agent's output"

Good reads and insightful comments thanks for sharing. I don't think I've ever actually used reduce() 

I'll submit an update.

**Review by Pwuts — COMMENTED** [review]
Does the way we *save* to the ai_config.yaml need to be fixed as well? Putting some quotes around all the text items?

**Comment by pr-0f3t** [comment]
> Does the way we _save_ to the ai_config.yaml need to be fixed as well? Putting some quotes around all the text items?

It is a possibility. I'll take a look.

However, the prompting sub-community is directly editing the ai_settings.yaml file. Many of them are opting for a static ai_settings.yaml and providing the actual instructions in a separate file that GPT-4 can directly parse: [1](https://github.com/samuelbutler/wonda), [2](https://github.com/psvensson/prompting/blob/main/coding/clog/program.txt). 

Spitballing: we could move the goals from AI settings to long-term memory. Or we could keep any user-provided initial instructions as free-form prompts, which is the approach that these projects usually take.

**Comment by pr-0f3t** [comment]
> Does the way we _save_ to the ai_config.yaml need to be fixed as well? Putting some quotes around all the text items?

Confirmed. AIConfig.save()'s doing the right thing.

**Review by collijk — APPROVED** [review]

**Comment by collijk** [comment]
Did you run black on your repo?  Our linter is flagging a failure due to trailing whitespace in your test file

**Comment by pr-0f3t** [comment]
Found it manually. Black wasn't detecting it.Squashed.

**Review by collijk — APPROVED** [review]

**Pull request merged**